### PR TITLE
feat: leverage SentenceTransformers' query/passage specific prompts

### PR DESCRIPTION
**Pull request by dbuades**
Some embedding models require different prompts for encoding queries versus encoding passages. SentenceTransformers provides this functionality through two features:

1. A model may be initialized with a prompt dictionary via the [prompts](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer) argument.
2. When calling `model.encode()`, we can pass a [prompt_name](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode) argument. If the argument matches a key in the prompt dictionary, that prompt will be used automatically.

The main change in this PR is the modification of the logic in `model_encode.py` to leverage this functionality. To maintain compatibility with instruct models that currently use the task name as `prompt_name`, I added a new `prompt_type` argument and renamed `prompt_name` to `task_name`. The other change is that `encode_queries` and `encode_corpus` now specify `query` and `passage` as their respective `prompt_type` argument, which matches SentenceTransformers' default. As a side benefit, this allows instruct models to also have specific prompts for queries and passages. 

Previously, the way to achieve this was by creating wrappers for these models: `arctic_models.py`, `bge_models.py`, `mxbai_models.py`, and `e5_models.py`. While this worked correctly, it added complexity to maintain.

With the changes in this PR, those wrappers are no longer needed, as we can now specify the `prompts` argument directly in the loader. In the case of arctic_models, since they define the prompts in their [config_sentence_transformers.json](https://huggingface.co/Snowflake/snowflake-arctic-embed-m-v1.5/blob/main/config_sentence_transformers.json), we don't even need to modify the loader.

@KennethEnevoldsen, please let me know your thoughts. I'm happy to discuss any concerns you may have!

## Checklist

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`.

### CODE REVIEW

File: mteb/models/e5_models.py
Comment (KennethEnevoldsen) [inline mteb/models/e5_models.py:1]: I like the simplification here. Would love to ensure that they produce the same results. Can you run it on a few cases?

File: mteb/models/e5_models.py [hunk 1]
```diff
@@ -113,69 +108,27 @@
...
+        sentence_transformers_loader,
+        model_name="intfloat/multilingual-e5-small",
+        revision="fd1525a9fd15316a2d503bf26ab031a61d056e98",
+        prompts={"query": "query: ", "passage": "passage: "},
+    ),
     name="intfloat/multilingual-e5-small",
     languages=XLMR_LANGUAGES,
     open_source=True,
-    revision="e4ce9877abf3edfe10b0d82785e83bdcb973e22e",
+    revision="fd1525a9fd15316a2d503bf26ab031a61d056e98",
```
Comment (KennethEnevoldsen) [inline mteb/models/e5_models.py:121]: ?

**Review by KennethEnevoldsen — COMMENTED** [review]
Thanks for the PR - I appreciate the simplification on the model side that this provides. it does move a lot of complexity to a single encode function. I have added a few reviewers to get their opinions as well.

If we go about changing this we can always deprecate the encode_corpus, encode_queries, encode_conversations etc. And then remove them in the next major version.

@tomaarsen if you have the time I would love your opinion on this as it concerns the integration with sentence transformers.

File: mteb/models/e5_models.py (hunk 1 above)
Comment (dbuades) [reply mteb/models/e5_models.py:121]: For some reason, `e4ce987` no longer exists as a revision in the Hub: https://huggingface.co/intfloat/multilingual-e5-small/commits/main

File: mteb/models/e5_models.py [hunk 2]
```diff
@@ -184,16 +137,26 @@ def encode_corpus(
...
+        sentence_transformers_loader,
+        model_name="intfloat/multilingual-e5-large",
+        revision="ab10c1a7f42e74530fe7ae5be82e6d4f11a719eb",
+        prompts={"query": "query: ", "passage": "passage: "},
+    ),
     name="intfloat/multilingual-e5-large",
     languages=XLMR_LANGUAGES,
     open_source=True,
-    revision="4dc6d853a804b9c8886ede6dda8a073b7dc08a81",
+    revision="ab10c1a7f42e74530fe7ae5be82e6d4f11a719eb",
```
Comment (dbuades) [inline mteb/models/e5_models.py:149]: @KennethEnevoldsen Same thing here with `4dc6d85` : https://huggingface.co/intfloat/multilingual-e5-large/commits/main

File: mteb/models/e5_models.py
Comment (dbuades) [reply mteb/models/e5_models.py:1]: No problem, I'll run it on 'main' and on this PR and report back.

File: mteb/evaluation/evaluators/model_encode.py [hunk 3]
```diff
@@ -6,35 +6,56 @@
...
-            kwargs.pop("prompt_name")
-        elif prompt_name not in model.prompts:  # type: ignore
+            kwargs.pop("prompt_name", None)
+
+        if task_name and prompt_type and f"{task_name}-{ prompt_type}" in model.prompts:
+            kwargs["prompt_name"] = f"{task_name}-{ prompt_type}"
+        elif task_name and task_name in model.prompts:
+            kwargs["prompt_name"] = task_name
+        elif prompt_type and prompt_type in model.prompts:
+            kwargs["prompt_name"] = prompt_type
```
[the remark this answers is not in the document]
Comment (dbuades) [reply mteb/evaluation/evaluators/model_encode.py:49]: It definitely feels a bit hacky but since SentenceTransformers offer a lot of flexibility with the prompt dictionary, I think this is something that we could handle on a potential refactor of `encode_corpus` / `encode_queries` / `encode_conversations` and I'd be happy to help.

Of course, if it's implemented upstream in ST, our life becomes easier, so I'm all for that solution too if Tom deems it appropriate.

File: mteb/models/e5_models.py
Comment (dbuades) [reply mteb/models/e5_models.py:1]: @KennethEnevoldsen You can find the results for the code in this PR vs main in [this Google Drive folder](https://drive.google.com/drive/folders/1vxL63ZkcO1bHeiL9uArIQQQ3GzcoVbPu?usp=sharing)

I ran `BAAI/bge-base-en-v1.5:a5beb1e` and `intfloat/multilingual-e5-base:d13f1b2` on `SCIDOCS`, `SciFact` and `PublicHealthQA`.

**Comment by orionw** [comment]
Really great PR, thanks @dbuades! Definitely needed and a great addition, although it is becoming complex. I am curious what @tomaarsen thinks as well.

My only other comments are just:
1. I think given the number of parameters we have now it'd be great to have some documentation (doesn't have to be from this PR, but MTEB should add some at some point)
2. I myself am not sure if there is any easy way to do this if I was only using one dataset.  Do I just pass a dict of something like {"my_dataset": "my prompt"}?

**Comment by dbuades** [comment]
> Really great PR, thanks @dbuades! Definitely needed and a great addition, although it is becoming complex. I am curious what @tomaarsen thinks as well.
> 
> My only other comments are just:
> 
> 1. I think given the number of flags we have now it'd be great to have some documentation (doesn't have to be from this PR, but MTEB should add some at some point)
> 2. I myself am not sure if there is any easy way to do this if I was only using one dataset.  Do I just pass a dict of something like {"my_dataset": "my prompt"}?

@orionw Thanks for the comments!
1. I agree that documentation would be required, I'll be happy to add it once we agree on a way to move forward with this proposal.
2. For now, we keep the `task_name` as an input so it's compatible with the current `task_to_instruction()` method. This means that we don't need to pass any dictionary if the dataset name is already contained in `TASKNAME2INSTRUCTIONS`. With a further refactor we could also simplify this logic.

**Comment by KennethEnevoldsen** [comment]
Thanks for the feedback @orionw - seems like we both think this is a reasonable change.

(@Muennighoff, @vaibhavad just making sure you guys are aware of this change as well) 

@dbuades let us move forward with this. Can I ask you to add the documentation as well as a deprecation warning for models using the  `encode_corpus` etc?

I would also like to make sure that you are aware of #1225 which I believe is compatible with this change (not that you need to integrate it)

File: mteb/models/arctic_models.py [hunk 4]
```diff
@@ -1,66 +0,0 @@
...
-
-
-arctic_m_v1_5 = ModelMeta(
-    loader=partial(ArcticWrapper, model_name="Snowflake/snowflake-arctic-embed-m-v1.5"),  # type: ignore
-    name="Snowflake/snowflake-arctic-embed-m-v1.5",
-    languages=["eng_Latn"],
-    open_source=True,
-    revision="97eab2e17fcb7ccb8bb94d6e547898fa1a6a0f47",
-    release_date="2024-07-08",  # initial commit of hf model.
-)
```
Comment (Muennighoff) [inline mteb/models/arctic_models.py:66]: Why is this one removed but kept for bge?

**Review by Muennighoff — COMMENTED** [review]
Looks great! Can we run 1-2 tasks for bge/mxbai/arctic/e5 to make sure results are the same before & after?

**Comment by KennethEnevoldsen** [comment]
> Looks great! Can we run 1-2 tasks for bge/mxbai/arctic/e5 to make sure results are the same before & after?

We have run bge and e5 (in resolved comment above), but I agree that it might be worth testing mxbai and arctic

File: mteb/models/arctic_models.py (hunk 4 above)
Comment (dbuades) [reply mteb/models/arctic_models.py:66]: @Muennighoff Good question. The reason is that arctic models already specify the `prompts` dictionary on the [config_sentence_transformers.json](https://huggingface.co/Snowflake/snowflake-arctic-embed-m-v1.5/blob/main/config_sentence_transformers.json) so we don't need to specify the prompts in the loader as with BGE and the other models.

I'll open PRs on BGE, e5 and mxbai model repos in the Hub proposing to add the prompt dictionary directly over there. If they are approved, we will be able to remove the specific files for other models too.

File: mteb/models/e5_models.py
Comment (dbuades) [reply mteb/models/e5_models.py:1]: @KennethEnevoldsen @Muennighoff I have updated the folder to include `Snowflake/snowflake-arctic-embed-m-v1.5` and `mixedbread-ai/mxbai-embed-large-v1`. I had errors running the arctic models in main, but you can see that the results from the model with this branch match those that are currently in the leaderboard for `SCIDOCS` and `SciFact`.

**Review by Muennighoff — APPROVED** [review]
amazing work on this!

**Comment by dbuades** [comment]
> Thanks for the feedback @orionw - seems like we both think this is a reasonable change.
> 
> (@Muennighoff, @vaibhavad just making sure you guys are aware of this change as well)
> 
> @dbuades let us move forward with this. Can I ask you to add the documentation as well as a deprecation warning for models using the `encode_corpus` etc?
> 
> I would also like to make sure that you are aware of #1225 which I believe is compatible with this change (not that you need to integrate it)

@KennethEnevoldsen Perfect! I have run the other models and added documentation on how to leverage the `prompts` dictionary with Sentence Transformers models.

Regarding the deprecation warnings, could you point me out where you would like me to add them please? 

The only other model wrapper that uses Sentence Transformers is `nomic_models.py` so, if they are only for that one, I can directly open a separate PR adapting it.

If, instead, we want to deprecate those methods (encode_corpus, encode_queries, encode_conversations) across all models and not only for Sentence Transformers, I can open a draft PR and work on adapting `DRESModel` and all the other model wrappers (open_ai, cohere, vertex, GritLM, etc.) and take the conversation there. What do you think?

**Comment by tomaarsen** [comment]
Hello!

Apologies for the radio silence, both here and on https://github.com/UKPLab/sentence-transformers/issues/2776. I hadn't yet responded on that issue because I'm a little conflicted on it - it can help simplify retrieval workflows, but it might give the impression that Sentence Transformers is for retrieval only. It also might be tricky to implement, given that every model has a different way/prompt it wants for queries/documents, and sometimes even has multiple options.

One of the potential implementations that I considered is roughly what you seem to be proposing here: using a prompt with the "query" and "passage" prompt_name's if they exist. Another consideration is leaving this to the third-party modules that integrate Sentence Transformers, as they might have different needs (e.g. what if BERTopic wants to use a "clustering", or SetFit wants a "classification" one. Should Sentence Transformers implement encode_clustering & encode_classification?)

In short:
1. I'm unsure whether Sentence Transformers will introduce abstractions such as encode_queries / encode_passages
2. If it does, it will likely involve the prompts/prompt_name, i.e. it'll rely on prompts with a specific key like "query", so encode_queries would probably be roughly equivalent to:
```python
def encode_queries(self, *args, **kwargs):
    try:
	    return self.encode(*args, **kwargs, prompt_name="query")
	except ValueError:
		logger.warning("The model did not contain a prompt with the \"query\" prompt_name.")
		return self.encode(*args, **kwargs)
```

So I like the simplifications that you're making here. 

- Tom Aarsen

**Comment by dbuades** [comment]
> Hello!
> 
> Apologies for the radio silence, both here and on [UKPLab/sentence-transformers#2776](https://github.com/UKPLab/sentence-transformers/issues/2776). I hadn't yet responded on that issue because I'm a little conflicted on it - it can help simplify retrieval workflows, but it might give the impression that Sentence Transformers is for retrieval only. It also might be tricky to implement, given that every model has a different way/prompt it wants for queries/documents, and sometimes even has multiple options.
> 
> One of the potential implementations that I considered is roughly what you seem to be proposing here: using a prompt with the "query" and "passage" prompt_name's if they exist. Another consideration is leaving this to the third-party modules that integrate Sentence Transformers, as they might have different needs (e.g. what if BERTopic wants to use a "clustering", or SetFit wants a "classification"? Should Sentence Transformers then also implement encode_clustering & encode_classification?)
> 
> In short:
> 
> 1. I'm unsure whether Sentence Transformers will introduce abstractions such as encode_queries / encode_passages
> 2. If it does, it will likely involve the prompts/prompt_name, i.e. it'll rely on prompts with a specific key like "query", so encode_queries would probably be roughly equivalent to:
> 
> ```python
> def encode_queries(self, *args, **kwargs):
>     try:
> 	    return self.encode(*args, **kwargs, prompt_name="query")
> 	except ValueError:
> 		logger.warning("The model did not contain a prompt with the \"query\" prompt_name.")
> 		return self.encode(*args, **kwargs)
> ```
> 
> So I like the simplifications that you're making here.
> 
> * Tom Aarsen

Thank you for taking the time to answer, Tom!

**Comment by KennethEnevoldsen** [comment]
> The only other model wrapper that uses Sentence Transformers is nomic_models.py so, if they are only for that one, I can open a separate PR to adapt it directly.

> If, instead, we want to deprecate those methods (encode_corpus, encode_queries, encode_conversations) across all models and not only for Sentence Transformers, I can open a draft PR and work on adapting DRESModel and all the other model wrappers (open_ai, cohere, vertex, GritLM, etc.) and take the conversation there. What do you think?

I think it is fine to do it in another PR and then get this one merged in. I believe we want to deprecate them fully but we will need to add the `promptType.conversation` for that to work. 

Sorry for having been a bit slow to adapt to this PR, we are submitting a paper for the multilingual MTEB benchmark in the coming week.

**Review by KennethEnevoldsen — APPROVED** [review]

**Comment by dbuades** [comment]
> > The only other model wrapper that uses Sentence Transformers is nomic_models.py so, if they are only for that one, I can open a separate PR to adapt it directly.
> 
> 
> 
> > If, instead, we want to deprecate those methods (encode_corpus, encode_queries, encode_conversations) across all models and not only for Sentence Transformers, I can open a draft PR and work on adapting DRESModel and all the other model wrappers (open_ai, cohere, vertex, GritLM, etc.) and take the conversation there. What do you think?
> 
> 
> 
> I think it is fine to do it in another PR and then get this one merged in. I believe we want to deprecate them fully but we will need to add the `promptType.conversation` for that to work. 
> 
> 
> 
> Sorry for having been a bit slow to adapt to this PR, we are submitting a paper for the multilingual MTEB benchmark in the coming week.  
> 
> 
> 
> Will just double check with you that this is ready to merge (couldn't find any outstanding issues that can't be tackled in future PRs

No worries at all, thank you for the review.

I'll start working on a draft PR with the proposed changes for DRESModel and the other wrappers next week.

Regarding this PR, all is ready, you can merge it when you have the time.

File: mteb/models/arctic_models.py (hunk 4 above)
Comment (Muennighoff) [reply mteb/models/arctic_models.py:66]: Actlly @KennethEnevoldsen do we generally want to keep models even if they don't require a loader? Afaik we are doing that for some models in `models/sentence_transformers_models.py` so should we also keep arctic just without the loader but for the metadata?

Comment (KennethEnevoldsen) [reply mteb/models/arctic_models.py:66]: Yes we should. (see also #1234). Generally @x-tabdeveloping and I have discussed moving all model metadata from the leaderboard repo to this repo. Keeping it all in one place.

**Comment by KennethEnevoldsen** [comment]
Perfect @dbuades we will need to add back the arctic model (which can be added in the sentence transformers file) as @Muennighoff noticed, but otherwise, we can merge (if you don't have time I will do this Wednesday and merge it in)

**Comment by dbuades** [comment]
> Perfect @dbuades we will need to add back the arctic model (which can be added in the sentence transformers file) as @Muennighoff noticed, but otherwise, we can merge (if you don't have time I will do this Wednesday and merge it in)

@KennethEnevoldsen No problem, I added it back 👌

**Comment by KennethEnevoldsen** [comment]
Perfect merging this in now

**Comment by dbuades** [comment]
By the way, I didn't forgot about the follow up PR, I've just been very busy for these past few weeks.
I will have time to get back to it soon 🤞 

FYI @KennethEnevoldsen

**Comment by Samoed** [comment]
@dbuades I've implemented some improvements in #1278 based on your idea. It would be great to get your review
